baseapp/baseapp.go

```diff
@@ -653,9 +661,22 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte) (gInfo sdk.GasInfo, re
 		anteEvents = events.ToABCIEvents()
 	}
 
+	if mode == runTxModeCheck {
+		err = app.mempool.Insert(ctx, tx.(mempool.Tx))
+		if err != nil {
+			return gInfo, nil, anteEvents, priority, err
+		}
+	} else if mode == runTxModeDeliver {
+		err = app.mempool.Remove(tx.(mempool.Tx))
```

Comment:
What happens if another node gossips a tx _after_ it is delivered (and removed from `app.mempool`) ? Tendermint has a cache mechanism that, while not perfect, helps with this use case.
Will you be relying on Tendermint's cache mechanism, or will you handle this case otherwise?

Reply:
@sergio-mena we're not relying or assuming anything about Tendermint here.

IIRC, Tendermint reaps the txs from its mempool prior to making a block proposal. Thus, the txs should not be in its mempool and thus gossiped.

Reply:
Two things:
1. "Reap" is probably a bad name for `ReapMaxBytesMaxGas` as the transactions are not _removed_ from the mempool when called a proposal time... it wouldn't make sense as we don't know if that proposal will be decided
2. Anyway, this is not the case I had in mind when I wrote this comment, but the following. If a node is a few heights late (and so it hasn't committed the block the transaction is part of) it's perfectly OK for that node to have that transaction in its mempool (I mean Tendermint's mempool) and still gossip it until it reaches the height that commits the transaction.